clippy_lints/src/erasing_op.rs

```diff
@@ -0,0 +1,62 @@
+use consts::{constant_simple, Constant};
+use rustc::hir::*;
+use rustc::lint::*;
+use syntax::codemap::Span;
+use utils::{in_macro, span_lint};
+
+/// **What it does:** Checks for erasing operations, e.g. `x * 0`.
+///
+/// **Why is this bad?** The whole expression can be replaced by zero.
+/// Most likely mistake was made and code should be reviewed or simplified.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// 0 / x; 0 * x; x & 0
+/// ```
+declare_lint! {
+    pub ERASING_OP,
+    Warn,
+    "using erasing operations, e.g. `x * 0` or `y & 0`"
+}
+
+#[derive(Copy, Clone)]
+pub struct ErasingOp;
+
+impl LintPass for ErasingOp {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(ERASING_OP)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ErasingOp {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, e: &'tcx Expr) {
+        if in_macro(e.span) {
+            return;
+        }
+        if let ExprBinary(ref cmp, ref left, ref right) = e.node {
+            match cmp.node {
+                BiMul | BiBitAnd => {
+                    check(cx, left, e.span);
+                    check(cx, right, e.span);
+                },
+                BiDiv => check(cx, left, e.span),
+                _ => (),
+            }
+        }
+    }
+}
+
+fn check(cx: &LateContext, e: &Expr, span: Span) {
+    if let Some(Constant::Int(v)) = constant_simple(cx, e) {
+        if v.to_u128_unchecked() == 0 {
+            span_lint(
+                cx,
+                ERASING_OP,
+                span,
+                "the operation is ineffective. Consider reducing it to `0`",
```

Comment:
I don't quite like our message here, as it implies the zero would be correct, where this will most likely not be the case. I'd rather warn "this operation will always return zero. This is likely not the intended outcome."

Reply:
Thanks, fixed